File: manual/code/generate_strategies/IR(summary)/cpp/cpp/function_bitarray_invert.cpp
#include <string>
#include <boost/dynamic_bitset.hpp>
#include <stdexcept>
class InvertBitsFunction {
public:
    std::string invert_bitarray(const std::string& bits, const std::string& start = "-1", const std::string& stop = "-1") {
        
        if (bits.empty() || bits.find_first_not_of("01") != std::string::npos) {
            throw std::invalid_argument("Input must be a string containing only '0' and '1'.");
        }

        
        int start_idx, stop_idx;
        try {
            size_t pos;
            start_idx = std::stoi(start, &pos);
            if (pos != start.length()) {
                throw std::invalid_argument("Start index must be an integer.");
            }
            
            stop_idx = std::stoi(stop, &pos);
            if (pos != stop.length()) {
                throw std::invalid_argument("Stop index must be an integer.");
            }
        } catch (const std::exception&) {
            throw std::invalid_argument("Invalid start or stop index format.");
        }

        
        if (start_idx != -1 && stop_idx != -1) {
            if (start_idx < 0 || stop_idx > bits.length() || start_idx >= stop_idx) {
                throw std::invalid_argument("Invalid start and stop indices.");
            }
        }

        
        boost::dynamic_bitset<> bit_array(bits.length());
        for (size_t i = 0; i < bits.length(); ++i) {
            bit_array[bits.length() - 1 - i] = (bits[i] == '1');
        }

        
        if (start_idx == -1 && stop_idx == -1) {
            bit_array.flip();
        } else {
            for (int i = start_idx; i < stop_idx; ++i) {
                bit_array.flip(bits.length() - 1 - i);
            }
        }

        
        std::string result;
        for (size_t i = 0; i < bit_array.size(); ++i) {
            result = (bit_array[i] ? '1' : '0') + result;
        }

        return result;
    }
};

File: data/python/src/function_bitarray_invert.py
from bitarray import bitarray

class InvertBitsFunction:
    def invert_bitarray(self, bits: str, start: int = None, stop: int = None) -> str:
        if not isinstance(bits, str) or not all(c in '01' for c in bits):
            raise TypeError("Input must be a string containing only '0' and '1'.")
        if start is not None and not isinstance(start, int):
            raise TypeError("Start index must be an integer or None.")
        if stop is not None and not isinstance(stop, int):
            raise TypeError("Stop index must be an integer or None.")
        if start is not None and stop is not None and (start < 0 or stop > len(bits) or start >= stop):
            raise ValueError("Invalid start and stop indices.")

        bit_array = bitarray(bits)

        if start is None and stop is None:
            bit_array.invert()
        else:
            for i in range(start, stop):
                bit_array[i] = not bit_array[i]

        return bit_array.to01()
